xcq-extension/procedural/src/decl_extensions.rs

```diff
@@ -0,0 +1,203 @@
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use std::hash::{Hash, Hasher};
+use syn::{
+    parse_macro_input, parse_quote, parse_str, punctuated::Punctuated, spanned::Spanned, token::Comma, Error, ExprCall,
+    Field, FnArg, Ident, ItemEnum, ItemImpl, ItemTrait, Pat, Result, TraitItem, Type, Variant,
+};
+
+use crate::utils::{generate_crate_access, generate_mod_name_for_trait};
+
+pub fn decl_extensions_impl(input: TokenStream) -> TokenStream {
+    let item_trait = parse_macro_input!(input as ItemTrait);
+    decl_extension_inner(&item_trait)
+        .unwrap_or_else(|e| e.to_compile_error())
+        .into()
+}
+
+pub fn decl_extension_inner(item_trait: &ItemTrait) -> Result<TokenStream2> {
+    let mod_name = generate_mod_name_for_trait(&item_trait.ident);
+
+    // Assume single config associated type.
+    let has_config = item_trait
+        .items
+        .iter()
+        .any(|item| matches!(item, syn::TraitItem::Type(_)));
+    let methods = methods(&item_trait.items)?;
+
+    let call_enum_def = call_enum_def(&item_trait.ident, &methods)?;
+    let dispatchable_impl = dispatchable_impl(&item_trait.ident, &methods)?;
+    let extension_id_impl = extension_id_impl(&item_trait.ident, &item_trait.items)?;
+
+    let runtime_metadata = crate::runtime_metadata::generate_decl_metadata(item_trait, has_config)?;
+
+    let expanded = quote! {
+        #item_trait
+        #[doc(hidden)]
+        #[allow(dead_code)]
+        pub mod #mod_name {
+            pub use super::*;
+            #call_enum_def
+            #dispatchable_impl
+            #extension_id_impl
+            #runtime_metadata
+        }
+        pub use #mod_name::*;
+    };
+    Ok(expanded)
+}
+#[derive(Clone)]
+struct Method {
+    /// Function name
+    pub name: Ident,
+    /// Information on args: `(name, type)`
+    pub args: Vec<(Ident, Box<Type>)>,
+}
+
+fn call_enum_def(trait_ident: &Ident, methods: &[Method]) -> Result<ItemEnum> {
+    let mut variants = Punctuated::<Variant, Comma>::new();
+    for method in methods {
+        let name = &method.name;
+        let mut args = Punctuated::<Field, Comma>::new();
+        for (name, ty) in &method.args {
+            let ty = replace_self_to_impl(ty)?;
+            args.push(parse_quote! {
+                #name: #ty
+            });
+        }
+        variants.push(parse_quote! {
+            #[allow(non_camel_case_types)]
+            #name {
+                #args
+            }
+        });
+    }
+    // Add phantom data
+    variants.push(parse_quote!(
+        #[doc(hidden)]
+        __Phantom(core::marker::PhantomData<Impl>)
+    ));
+    Ok(parse_quote!(
+        #[derive(parity_scale_codec::Decode)]
```

Comment:
will this work if `parity_scale_codec` was imported with a different name? maybe we need to do something like `#xcq_extension::deps::parity_scale_codec::Decode`